org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java

```diff
@@ -1280,4 +1283,39 @@ public MavenExecutionContext getExecutionContext() {
     return MavenExecutionContext.getThreadContext();
   }
 
+  /**
+   * Execute the given {@link MavenExecutionRequest} in the context of m2eclipse and could be seen as an embedded run of
+   * the maven cli, at laest it tries to replicate as much as possible from that.
+   */
+  @Override
+  public MavenExecutionResult execute(MavenExecutionRequest request) {
+    MavenExecutionContext context = new MavenExecutionContext(this, request);
+    try {
+      return context.execute((innerContext, monitor) -> {
+
+        //First look up required components...
+        MavenExecutionRequestPopulator requestPopulator = lookup(MavenExecutionRequestPopulator.class);
+        EventSpyDispatcher eventSpyDispatcher = lookup(EventSpyDispatcher.class);
+        Maven maven = lookup(Maven.class);
+        //populate the defaults (if not already given)...
+        try {
+          requestPopulator.populateDefaults(request);
```

Comment:
I have considered only how maven (cli) behaves here, that is you create a execution request, execute it and get a result, there is no such thing as "joining" an execution and I'm not sure if it is even useful here.

Reply:
> `IMaven/MavenImpl` already has a method `populateDefaults()` that is doing exactly this.

Yes thats why I wrote I have wondered why this was/is called at all at that place of code, but as maven-cli is also doing it as part of the setup I think we can safely replicate this here, but I won't call the `populateDefaults` here as it unnecessarily wraps the exception we might better handle here directly.

Reply:
> However when I look at the implementation of `MavenExecutionContext.execute(MavenProject, ICallable, IProgressMonitor)` then I wonder if the mentioned methods are called any way.
> 

I wanted more to say that `populateDefaults()` is already called within `execute()` (what is called here) before the lambdas body is called. So it looks like `populateDefaults()` is just called a second time here.


> > `IMaven/MavenImpl` already has a method `populateDefaults()` that is doing exactly this.
> 
> Yes thats why I wrote I have wondered why this was/is called at all at that place of code, but as maven-cli is also doing it as part of the setup I think we can safely replicate this here, but I won't call the `populateDefaults` here as it unnecessarily wraps the exception we might better handle here directly.

It is definitely good to use the Maven implementation as guidance but I think since this is not exactly Maven-CLI we should adapt the code if necessary/suitable.

Not wrapping the execution is a good point, but I suspect that the exception will already be thrown before when `populateDefaults()` executed the first time in `MavenExecutionContext.execute()`.

Reply:
> I wanted more to say that `populateDefaults()` is already called within `execute()` (what is called here) before the lambdas body is called. So it looks like `populateDefaults()` is just called a second time here.

possibly yes, but I don't think this harms much, we probably could find a way later to avoid that. Interestingly I thin the method might better be (package) private than part of the IMaven API then (it is only refrecned on two places anyways).

> Not wrapping the execution is a good point, but I suspect that the exception will already be thrown before when `populateDefaults()` executed the first time in `MavenExecutionContext.execute()`.

At least there are branches where `populateDefaults()` is not executed and I won't mid the duplicate call here as it is only populates the defaults if not already set and don't want to change to much at once... especially because of the calls site I replaced also explicitly called that component.